### ISSUE TITLE
Access violation in `nth_packed_object_offset`

### ISSUE THREAD

**Issue by csware** [libgit2/libgit2#3556]
Try to read the "/Cross/" tree of commit "a9010e2d805a424e2c9279da22b56b9b1eba9b7c" (7th entry of the root tree) and when executing "git_tree_entry_to_object" on the repository you can get from https://yadi.sk/d/50mlHdydkLmZn and you'll get an access violation.

	libgit2_tgit.dll!nth_packed_object_offset(const git_pack_file * p, unsigned int n) Zeile 1190	C
 	libgit2_tgit.dll!pack_entry_find_offset(__int64 * offset_out, git_oid * found_oid, git_pack_file * p, const git_oid * short_oid, unsigned __int64 len) Zeile 1340	C
 	libgit2_tgit.dll!git_pack_entry_find(git_pack_entry * e, git_pack_file * p, const git_oid * short_oid, unsigned __int64 len) Zeile 1374	C
 	libgit2_tgit.dll!pack_entry_find_inner(git_pack_entry * e, pack_backend * backend, const git_oid * oid, git_pack_file * last_found) Zeile 247	C
 	libgit2_tgit.dll!pack_entry_find(git_pack_entry * e, pack_backend * backend, const git_oid * oid) Zeile 264	C
 	libgit2_tgit.dll!pack_backend__read(void * * buffer_p, unsigned __int64 * len_p, git_otype * type_p, git_odb_backend * backend, const git_oid * oid) Zeile 373	C
 	libgit2_tgit.dll!odb_read_1(git_odb_object * * out, git_odb * db, const git_oid * id, unsigned char only_refreshed) Zeile 846	C
 	libgit2_tgit.dll!git_odb_read(git_odb_object * * out, git_odb * db, const git_oid * id) Zeile 878	C
 	libgit2_tgit.dll!git_object_lookup_prefix(git_object * * object_out, git_repository * repo, const git_oid * id, unsigned __int64 len, git_otype type) Zeile 164	C
 	libgit2_tgit.dll!git_object_lookup(git_object * * object_out, git_repository * repo, const git_oid * id, git_otype type) Zeile 200	C
 	libgit2_tgit.dll!git_tree_entry_to_object(git_object * * object_out, git_repository * repo, const git_tree_entry * entry) Zeile 317	C

Line numbers (Zeile) according to 6aa06b65cfcce164d7ad3ef184f40acb68406078.

`git log` outputs the following error:

    error: packfile .git/objects/pack/pack-b17748a8c9805855e5cf2e5958f90f5282b4ec7a.pack does not match index
    error: packfile .git/objects/pack/pack-b17748a8c9805855e5cf2e5958f90f5282b4ec7a.pack does not match index
    error: Could not read 5b9ce45660f01f87f17ed2211a1c01540d903a41
    fatal: Failed to traverse parents of commit d54288eb977c475e5eb044373c07e7873a1d90e7

Yes, the repository is broken, but libgit2 should not crash.

**Comment by csware** [comment]
See https://github.com/csware/libgit2/commit/8f022256c5d6e548434ed6b8142919a555c6ab1a for a fix, but I doubt that this fix is comprehensive as there might be a lot of places where proper bound checks are missing as user provided data is trusted without proper validation this might also be a security problem.

**Comment by carlosmn** [comment]
From that commit, it looks like we'd end up asking for an object in the index at an offset past the end. But just before we call `nth_packed_object_offset()` we have checks to see if `pos` is less than the number of objects. If the check is not detecting a bad case, it looks like that's what should get fixed to detect this case.

**Comment by csware** [comment]
There are also dereferecnes in `pack_entry_find_offset` which might also need to be checked before performing them.

**Comment by carlosmn** [comment]
I just tried to list this tree with git itself (the actual tree is `c6df1165675344c3cb5718965d3f9b68ec3be992` ftr).

```
% git cat-file -p c6df1165675344c3cb5718965d3f9b68ec3be992           
zsh: segmentation fault (core dumped)  git cat-file -p c6df1165675344c3cb5718965d3f9b68ec3be992
```

which suggests that the actual index contains bad information. Which makes this a different issue: one of protecting against bad input.

/cc @peff

**Comment by peff** [comment]
Yes, the `.idx` is corrupt about 24K in. The packfile is fine, so no data is lost. You can do:

```
$ pack=$(git fsck 2>&1 | head -1 | perl -ne '/pack-.{40}/ and print $&')
$ echo pack
pack-b17748a8c9805855e5cf2e5958f90f5282b4ec7a

$ cd .git/objects/pack
$ mkdir broken
$ mv $pack.* broken
$ git index-pack --stdin <broken/$pack.pack
pack    b17748a8c9805855e5cf2e5958f90f5282b4ec7a
$ cmp $pack.idx broken/$pack.idx
pack-b17748a8c9805855e5cf2e5958f90f5282b4ec7a.idx broken/pack-b17748a8c9805855e5cf2e5958f90f5282b4ec7a.idx differ: byte 24577, line 85
```

which recovers the pack and shows the breakage.  You can `cmp -l` to see the details, but after that byte, it's pretty screwed. The breakage starts right after a 4K boundary, so the plausible guess is filesystem corruption.

> Which makes this a different issue: one of protecting against bad input.

Yeah, we actually discussed this upstream about a month ago, but it looks like Junio never followed through on [his patch](http://thread.gmane.org/gmane.comp.version-control.git/283377/focus=283500).

It's less dire than you might think, though. The `.idx` files are always locally generated, so an attack would require convincing somebody to read your repository files directly (e.g., by sending them a link to a broken repo to investigate :trollface:).

The likely attack surface is multi-user shared servers, but I don't think there's any vulnerability for most users (it's also a bounds-check failure on memory which is meant to be read-only, meaning the worst you can do is read random memory; I doubt this could be turned into remote code execution).

**Comment by carlosmn** [comment]
This would presumably be detectable if we check the tailing hash for the index, right? How come git doesn't (we copied this code, so we'd share a rationale, presumably)? The perf aspect likely wouldn't be as bad in libgit2 since we can amortize the cost of hashing over the multiple operations you do.

But an intentionally-crafted idx would be different... we would then still have to do some bounds checking.

**Comment by peff** [comment]
Yes, the trailing hash does show the corruption:

```
verify_hash () {
	size=$(stat --format=%s $1) &&
	head -c $(($size - 20)) $1 | openssl sha1 -binary >actual &&
	tail -c 20 $1 >expect &&
	cmp expect actual &&
	echo OK
}

$ verify_hash pack-a7908f89409f25f9d356cf0b92b06f606099601b.idx
OK
$ verify_hash pack-b17748a8c9805855e5cf2e5958f90f5282b4ec7a.idx
expect actual differ: byte 1, line 1
```

You _should_ be able to use `git verify-pack` to check this, but it segfaults before finishing the data, due to the out-of-bounds access. :trollface:

When opening a `.idx`, git checks that the sha1 of the `.pack` file mentioned in the header of the `.idx` matches what is in the trailer of the `.pack`. But it doesn't re-compute the trailer sha1 of either file during normal operation; it's just too expensive for too little gain.

So during normal operation, we have to assume that there can be corruption we don't know about and protect our accesses (which of course is no different than a `.idx` that is simply malicious in the first place, and whose sha1 checksums the malicious bits).

**Comment by carlosmn** [comment]
Hashing the packfiles is definitely too expensive, though the `.idx` might not be terrible for a long-lived process.

But I don't actually see the `.idx` packfile trailer match the packfile itself.

```
00007d80: 0d77 8001 ff00 d683 f801 750a 6828 0024  .w........u.h(.$
00007d90: 9d6e e8bd 9cff ff02 6841 20ff 757c e89c  .n......hA .u|..
00007da0: 1d40 0000 85c0 740c 4375 0304 e962 c0eb  .@....t.Cu...b..
00007db0: 837d 3800                                .}8.
```

The packfile trailer starts with the last four bytes (two groups) on the first line, and if we did check, we should be able to tell that this `.idx` is corrupt, even without checking the hash of the file content itself.

We don't do this in libgit2, which means any check in git would be somewhat new. But failure to recognise this as corrupt leads me to believe git does not in fact check this either.

**Comment by peff** [comment]
Hmph, you're right. I was thinking the `.idx`'s pack sha1 was at the front, but it's at the very end, so it is corrupted. Git _does_ notice and refuse to look at the objects in the pack:

```
$ git rev-list --all
a9010e2d805a424e2c9279da22b56b9b1eba9b7c
7195e5da55a40d15e29ad80562668bdd6895441f
de085464faa417eecb00b637691d5b1d599cef1d
ff27773b39c4456f8690ced819d0c54ce5c512ab
cf1be84ef57635f6ff1f2fe3b27d4604c0e258de
error: packfile /home/peff/perl_gcc/.git/objects/pack/pack-b17748a8c9805855e5cf2e5958f90f5282b4ec7a.pack does not match index
error: packfile /home/peff/perl_gcc/.git/objects/pack/pack-b17748a8c9805855e5cf2e5958f90f5282b4ec7a.pack does not match index
error: Could not read 5b9ce45660f01f87f17ed2211a1c01540d903a41
fatal: Failed to traverse parents of commit d54288eb977c475e5eb044373c07e7873a1d90e7
```

I think the `cat-file` you showed is probably the exception; there are some speed hacks in `cat-file`, though I'm surprised it bypasses the usual pack opening.

**Comment by peff** [comment]
> I think the cat-file you showed is probably the exception; there are some speed hacks in cat-file, though I'm surprised it bypasses the usual pack opening.

Ah, I see. We look up in the `.idx` first, and only if we find something do we open the pack and check that it's valid, matches the index, etc. So we may do the object lookup on a bogus `.idx` before realizing it is totally broken.

I _think_ that's fine. It's faster (if we never find an object in an index, we don't have to open the pack at all), and as long as the lookup code is properly defensive to avoid segfaulting (which it should be regardless), there's no harm done if we find out immediately after that our pack is bogus.

As for why it triggers with `cat-file` and not `rev-list`, I think `rev-list` may simply get lucky that it doesn't segfault for declaring the pack bogus (because we loaded some _other_ packs first).

**Comment by carlosmn** [comment]
That makes sense. We might also just be looking at a correct part of the `.idx` before via the rev-list so we just end up actually performing the check.

I just checked with libgit2, and we do error out (though apparently without setting an error message, which should be fixed). So I figure this should be enough to protect against bad `.idx` files since we do perform a check.

**Comment by peff** [comment]
Yeah, I think that's more plausible. We hit some other object in the idx first, it doesn't segfault, then we notice the bogus match with the `.pack`.

FYI, I just sent [two fixes](http://thread.gmane.org/gmane.comp.version-control.git/287390) upstream to git. The first is what we've been discussing here, but I also noticed a problem in `use_pack` that libgit2 might have inherited.

**Comment by carlosmn** [comment]
Fixed via #3630
